earth_enterprise/src/common/FileAccessorPluginLoader.cpp

```diff
@@ -0,0 +1,107 @@
+// Copyright 2020 The Open GEE Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "FileAccessorPluginLoader.h"
+#include "POSIXFileAccessor.h"
+#include "notify.h"
+
+void ScanPluginDirectory(const std::string &pluginDir, std::vector<std::string> &files) {
+    boost::filesystem::path p(pluginDir);
+    boost::system::error_code errorCode;
+    boost::filesystem::directory_iterator start(p, errorCode);
+    if (errorCode.value() != 0){
+        notify(NFY_WARN, "Unable to scan file accessor plugin directory, %s. %s.", pluginDir.c_str(), errorCode.message().c_str());
+        return;
+    }
+
+    boost::filesystem::directory_iterator end;
+    std::transform(start, end, std::back_inserter(files), 
+        [pluginDir](const boost::filesystem::directory_entry& entry) {
+            return pluginDir + entry.path().leaf().string(); });
+}
+
+FileAccessorPluginLoader& FileAccessorPluginLoader::Get() {
+    static FileAccessorPluginLoader loader;
+    return loader;
+}
+
+void FileAccessorPluginLoader::DefaultLoadPluginsImpl (const std::string &pluginDirectory, std::vector<PluginFactory> &factories) {
+    assert(factories.size() == 0);
+    
+    std::vector<std::string> files;
+    ScanPluginDirectory(pluginDirectory, files);
+    for (auto s : files) {
+        FileAccessorPluginLoader::PluginHandle handle = dlopen(s.c_str(), RTLD_NOW);
+        if (handle){
+            const std::string getFactoryFunctionName = "get_factory_v1";
+            get_factory_t get_factory = reinterpret_cast<get_factory_t>(dlsym(handle, getFactoryFunctionName.c_str()));
+            FileAccessorFactory *pFactory = get_factory != nullptr ? get_factory() : nullptr;
+            if(pFactory){
+                factories.push_back({handle,pFactory});
+            }
+            else{
+                // 's' was a valid shared library, but did not provide the required function
+                dlclose(handle);
+                notify(NFY_WARN, "File accessor plugin %s does not implement the expected function %s", s.c_str(), getFactoryFunctionName.c_str());
+            }
+        }
+        else {
+            // 's' was something other than a shared library
+            notify(NFY_WARN, "Unable to open file accessor plugin %s", dlerror());
+        }
+    }
+}
+
+FileAccessorPluginLoader::FileAccessorPluginLoader(
+        LoadPluginFunc LoadImpl /*= nullptr*/, 
```

Comment:
Minor: remove commented-out initialization

Reply:
I put that in there on purpose so anyone looking at the implementation would know right away that there are default values for those parameters. I'd prefer to leave it unless you have a strong preference to the contrary.

Reply:
Ah, OK. I see that they are set in the .h file. That makes sense. Still minor: Can you update it so that it looks like a text comment rather than commented out code? e.g. "default: nullptr"

Reply:
Good idea. Updated as described.